packages/wdio-browserstack-service/src/util.ts

```diff
@@ -97,7 +106,157 @@ export function getParentSuiteName(fullTitle: string, testSuiteTitle: string): s
+    const configWithoutPII = JSON.parse(JSON.stringify(userConfig));
+    ['user', 'username', 'key', 'accessKey'].forEach(key => delete configWithoutPII[key])
+    const finalServices = []
+    try {
+        for (const serviceArray of configWithoutPII.services) {
+            if (Array.isArray(serviceArray) && serviceArray.length >= 2 && serviceArray[0] === 'browserstack') {
+                for (let idx=1; idx<serviceArray.length; idx++) {
+                    ['user', 'username', 'key', 'accessKey'].forEach(key => delete serviceArray[idx][key])
+                }
+                finalServices.push(serviceArray)
+                break
+            }
+        }
+    } catch (err) {
+        /* Wrong configuration like strings instead of json objects could break this method, needs no action */
+    }
+    configWithoutPII.services = finalServices
+    return configWithoutPII
+}
+
+function setCredentialsForCrashReportUpload(options: BrowserstackConfig & Options.Testrunner, config: Options.Testrunner) {
+    credentialsForCrashReportUpload = {
+        username: getObservabilityUser(options, config),
+        password: getObservabilityKey(options, config)
+    }
+    process.env.CREDENTIALS_FOR_CRASH_REPORTING = JSON.stringify(credentialsForCrashReportUpload)
+}
+
+export function setConfigDetails(userConfig: Options.Testrunner, capabilities: Capabilities.RemoteCapability, options: BrowserstackConfig & Options.Testrunner) {
+    const configWithoutPII = filterPII(userConfig)
+    userConfigForReporting = {
+        framework: userConfig.framework,
+        services: configWithoutPII.services,
+        capabilities: capabilities,
+        env: {
+            'BROWSERSTACK_BUILD': process.env.BROWSERSTACK_BUILD,
+            'BROWSERSTACK_BUILD_NAME': process.env.BROWSERSTACK_BUILD_NAME,
+            'BUILD_TAG': process.env.BUILD_TAG
+        }
+    }
+    process.env.USER_CONFIG_FOR_REPORTING = JSON.stringify(userConfigForReporting)
+    setCredentialsForCrashReportUpload(options, userConfig)
+}
+
+function processError(error: any, fn: Function, args: any) {
+    log.error(`Error in executing ${fn.name} with args ${args}: ${error}`)
+    let argsString: string
+    try {
+        argsString = JSON.stringify(args)
+    } catch (e) {
+        argsString = util.inspect(args, { depth: 2 })
+    }
+    uploadCrashReport(`Error in executing ${fn.name} with args ${argsString} : ${error}`, error && error.stack)
+}
+
+function o11yErrorHandler(fn: Function) {
+    return function (...args: any) {
+        try {
+            // @ts-ignore
+            const result = fn(...args)
+            if (result instanceof Promise) {
+                return result.catch(error => processError(error, fn, args))
+            }
+            return result
+        } catch (error) {
+            processError(error, fn, args)
+        }
+    }
+}
+
+// https://tugayilik.medium.com/error-handling-via-try-catch-proxy-in-javascript-54116dbf783f
```

Comment:
Can you explain why we need this?

Reply:
This is a wrapper for error handling of methods in a class. Have [added comments](https://github.com/webdriverio/webdriverio/pull/10183/commits/e8539f4a450860e198b32d961dc13d55fdfb7658) in the code.